add validation logic for APIVersion fields of HPA

lalitc375 opened the pull request:
New validation logic follows the API ratcheting principle,  will not be executed for already stored invalid if the corresponding fields or item in array is not modified.



#### What type of PR is this?

/kind cleanup


#### What this PR does / why we need it:
This adds validation logic on APIVersion fields of HorizontalPodAutoscaler type. There are no validation on this field, Invalid API version can result in non functioning HPA. 

This also adds additional checks for ScaleTargetRef field, to make sure it is not empty when Kind is not replicationController.

#### Which issue(s) this PR is related to:

Fixes #60511 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
APIVersion fields of the HorizontalPodAutoscaler API are now validated to ensure created API objects function properly
```

lalitc375 commented:
/assign @cici37

lalitc375 commented:
/retest

lalitc375 commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -51,34 +53,46 @@ func ValidateScale(scale *autoscaling.Scale) field.ErrorList {
-	if refErrs := ValidateCrossVersionObjectReference(autoscaler.ScaleTargetRef, fldPath.Child("scaleTargetRef")); len(refErrs) > 0 {
+
+	validateAPIVersion := oldAutoScaler == nil || newAutoScaler.ScaleTargetRef.APIVersion != oldAutoScaler.ScaleTargetRef.APIVersion
+	gvValidation := func(gv schema.GroupVersion) error {
```
This make sense if we plumb other values in the struct. I didn't want to pass two boolean value to the method in the current implementation.

liggitt replied:
in general, plumbing opts as far down as possible is nice, since it keeps the boolean definitively named and documented, and means if we ever need more options we don't have to replumb

liggitt commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -51,34 +53,46 @@ func ValidateScale(scale *autoscaling.Scale) field.ErrorList {
 	}
-	if refErrs := ValidateCrossVersionObjectReference(autoscaler.ScaleTargetRef, fldPath.Child("scaleTargetRef")); len(refErrs) > 0 {
+
+	validateAPIVersion := oldAutoScaler == nil || newAutoScaler.ScaleTargetRef.APIVersion != oldAutoScaler.ScaleTargetRef.APIVersion
```
> Each member of this array, we have to make the ratcheting decision.

generally, we don't worry about per-item ratcheting ... we require new things to be valid on create, and don't worry about already-invalid things updating to be invalid in slightly different ways

lalitc375 replied:
This needs more discussion. We will discuss this more in the DV working group.  For now, Allowed transition from any invalid to any valid for metrics array.

lalitc375 committed:
add validation logic for APIVersion fields of HPA

New validation logic follows the API ratcheting principle,  will not be executed for already stored invalid if the corresponding fields or item in array is not modified. Please enter the commit message for your changes. Lines starting

liggitt reviewed (commented):
one tweak on the error message, then lgtm

liggitt commented:
lgtm, go ahead and squash to a single commit

lalitc375 commented:
> lgtm, go ahead and squash to a single commit

done
